### ISSUE TITLE
Audacity won’t export audio, claims avformat-60.dll is missing

### ISSUE THREAD

**Issue by pfurrie** [audacity/audacity#5743]
```
Program build date: Nov 16 2023
Commit Id: 9dd60f of Thu Nov 16 14:58:21 2023 +0300
Build type: CMake Release build (debug level 1), 64 bits
```

`FFMPEG: Installed from FFmpeg_5.0.0_for_Audacity_on_Windows_x86_64.exe`

Have recorded multiple projects. Audacity project files save fine and can be reloaded fine. Can hear audio within those projects. When I go to export the files, I do it with these settings:

![image](https://github.com/audacity/audacity-support/assets/11449721/b9e7cc89-489f-4f35-a29d-e3b9dcf45a02)

Should be a cake-walk. When the file is exported, a quick progress dialogue is displayed:

![image](https://github.com/audacity/audacity-support/assets/11449721/e40ef2ad-997a-448d-a714-9d7cd701f5f1)

This is on-screen for about two seconds. A file is created using the name I gave it:

![image](https://github.com/audacity/audacity-support/assets/11449721/5af01397-8f7f-4b6f-98a2-188cfc85ff39)

Open the file in VLC, the file is the correct amount of time (46:57) but no sound. Open the WAV file in Audacity, it has none of the project sound, just some light buzzing… none of the people talking.

Checked the log file:

```
20:13:58: Audacity 3.4.2
20:13:58: Error: Failed to load shared library ‘avformat-60.dll’ (error 126: The specified module could not be found.)
20:13:58: FFmpeg libraries loaded successfully from: C:\Program Files\FFmpeg For Audacity\avformat-59.dll
20:14:01: sqlite3 message: (1) no such table: project in “SELECT 1 FROM project LIMIT 1;”
20:14:01: Error: Failed to load shared library ‘avformat-60.dll’ (error 126: The specified module could not be found.)
20:14:01: FFmpeg libraries loaded successfully from: C:\Program Files\FFmpeg For Audacity\avformat-59.dll
20:14:09: Loaded 70 string 1.406250 Kb in size
20:14:09: Project loaded in 21 ms
20:14:36: Operation ‘Export’ took 1.400000 seconds. Poll was called 23 times and took 0.044959 seconds. Yield was called 15 times and took 0.028434 seconds.
```

Here’s the FFMPEG directory:

![image](https://github.com/audacity/audacity-support/assets/11449721/473aa9f2-bec7-48c7-bb49-39672de65c70)

To recap:

- Have the latest version of Audacity (3.4.2, 64 bit) and of FFMPEG for Audacity 64-bit.
- Saving a simple WAV file
- Audacity is asking for “avformat-60.dll” but it doesn’t exist.

Tried with MP3 output, which took 17 seconds to export but just had a bit louder buzzing sound with none of the actual project audio.

I’ve used Audacity elsewhere for years, never had this problem.
**What could be wrong? What is “avformat-60.dll” and why is it missing?**

**Comment by pfurrie** [comment]
Okay, I guess there was some guess-work there based on this line in the log that comes immediately after the error line:
`20:13:58: FFmpeg libraries loaded successfully from: C:\Program Files\FFmpeg For Audacity\avformat-59.dll`
Considering that the error was about "avformat-60.dll is missing" and that the ffmpeg file mentioned is "avformat-59.dll," I think it is a reasonable guess to imagine that the problem might be a missing ffmpeg library.

However, here's a screenshot of Audacity with the project open:
![image](https://github.com/audacity/audacity/assets/11449721/b78a096d-e51b-4411-88e8-85c263da2791)

This is super-vanilla.  Record audio and then export.  Only thing off-script is that I'm trying to same in mono, and to do that I just check the "mono" box in the export dialogue.  The file it exports is always the correct length when opened in VLC, but like I said before, there is no sound, and re-opening that exported WAV file in Audacity shows no audio.  

I tried twice to attach the project, but first found the AUP3 extension isn't allowed (!), and then zipped the file but it was too big (limited to 25 MB, and the project is 459 MB).

**Comment by crsib** [comment]
Hmm, my only guess is that you may have muted the track. I've checked exporting stereo to mono, and it worked just fine. @dozzzzer @chinakov can you please help, guys?

**Comment by dozzzzer** [comment]
I couldn't reproduce it in 3.4.2 as well as in the latest master branch on Windows 11 machine.
@pfurrie please consider sharing your project via Dropbox / Google Drive / OneDrive etc.

**Comment by pfurrie** [comment]
When I get home today I'll try that.  I'm the meantime, can you run a
search on your computer for avformat-60.dll?

**Comment by pfurrie** [comment]
@dozzzzer @chinakov @crsib

Here is the Google Drive link to the folder with a file of one of the projects that is exhibiting this problem (for me):
https://drive.google.com/drive/folders/17xLW0mg4WfjweeGLav4yZS_NlZ1cHlrj?usp=sharing

NOTE: I'm betting the problem isn't with the project but with either the software, expecting the avformat-60.dll.  Do any of you know just what **_is_** avformat-60.dll?  Audacity is throwing an error about that specific file that it apparently wants.

**Comment by pfurrie** [comment]
That's.  I've granted access

**Comment by crsib** [comment]
> That's.  I've granted access

Thank you! I am able to reproduce the issue on your project, we will take a look

**Comment by crsib** [comment]
So, I've been looking at what your data **looks** like and had a growing suspicion that your channels are in the counter phase. Quickly splitting the track and running "Mix and Render" has proven me right.

That's it - when you export as a mono file, channels in the project cancel out each other. If you want to export your data as a mono file, I can advise you to split your project into two mono tracks and mute or delete one of them before exporting. 

![image](https://github.com/audacity/audacity/assets/2660628/f78ad2da-c784-42e4-b9a9-4cd0a299d93f)

**Comment by crsib** [comment]
You can see that the peaks look inverted:
![image](https://github.com/audacity/audacity/assets/2660628/1da4a4e7-84f5-4020-bfb7-d8eea0a1ab9d)

And here is the analysis plugin running over your track; you can observe 180 degrees difference between the phases
![image](https://github.com/audacity/audacity/assets/2660628/cb237592-f94a-457f-a021-8e0fc2fb950a)

**Comment by JulianSMoore** [comment]
> I am sure that `avformat-60.dll` has nothing to do with WAV or MP3 export. At the start, Audacity tries to find the newest possible FFmpeg version, logging quite extensively so we can help users who have problems with FFmpeg. We should update how it logs so it is less confusing, but that won't be a priority now.

@crsib You mean, Audacity generates *possible* version numbers and then looks to see if they exist, so the failure to find "-60" implies nothing about whether it should have been found?

I'd like to suggest that it _should_ be a priority to make this MUCH clearer, because one only looks in the log when there's an issue, and any sensible person sees the first issue noted as the probable cause of whatever went wrong... and in this case then wastes hours, because AFAICT (and I've been looking!) no one has explained this, so all failures that have "avformat-60" not found end _eventually_ up just solving the *real* problem and the misleading info is overlooked/forgotten again and again.

**Comment by crsib** [comment]
The main goal for logs is to allow developers to identify the issue. We log and we will log errors of opening ffmpeg libraries so we can solve issues on user machines, such as architecture mismatch. The way it is logged now is far from perfect, but it does the job of identifying issues with ffmpeg for us.

**Comment by JulianSMoore** [comment]
Sure. Could you at least say whether my guess at how dll existence is being tested is correct? Then at least a definitive, findable in principle, answer would exist here.

**Comment by crsib** [comment]
I'm not sure which question do you want the answer to.

The idea is quite simple: we have a built-in support for several versions of FFmpeg libraries. I would say that right now we support avformat versions 60, 59, 58, 57 and 55. During the startup we try to loaded the newest version present on the computer. We look in PATH, under a specific registry key and in the folder specified by user in prefs.

We can't ship FFmpeg nor we know which version is present on the end user computer. For this reason we have this process of looking for the best version possible

**Comment by JulianSMoore** [comment]
OK, so the avformat search begins with the latest existing supported avformat dll; that makes sense. Then it falls back version by version, so the first failure is not necessarily an "error", but a "warning"/info.

An "error" would only actually arise if **no** suitable avformat dll were found. Wording change "Error" -> "Warning"/"Information" would seem like a simple fix except I bet the "Error" comes from the language construct trying/failing as it iterates through versions, which, if true, would mean a small code change.

For the benefit of all crawlers my summary would be: "avformat-60.dll not found solved"...

"Failed to load shared library" "avformat-60.dll" "error 126" not necessarily a "real" "error". If the log subsequently says "FFmpeg libraries loaded successfully" and references another avformat-xx.dll, then all is well.

Thanks for helping clear that up.
